### PR TITLE
Multi-GPU for DF and DFT

### PR DESCRIPTION
**Pull request by wxj6000**
- Distributed CDERI
- Multi-GPU for DF and DFT

### CODE REVIEW

File: gpu4pyscf/df/df_jk.py [hunk 1]
```diff
@@ -243,98 +241,101 @@ def to_cpu(self):
         obj = self.undo_df().to_cpu().density_fit()
         return utils.to_cpu(self, obj)
 
-def get_jk(dfobj, dms_tag, hermi=0, with_j=True, with_k=True, direct_scf_tol=1e-14, omega=None):
-    '''
-    get jk with density fitting
-    outputs and input are on the same device
-    TODO: separate into three cases: j only, k only, j and k
+def _jk_task_with_mo(dfobj, dms, mo_coeff, mo_occ, device_id,
```
Comment (sunqm) [inline gpu4pyscf/df/df_jk.py:244]: In this function and other helper functions, treat `device_id` as kwarg, and set its default value to 0

File: gpu4pyscf/lib/cupy_helper.py [hunk 2]
```diff
@@ -82,6 +83,51 @@ def get_avail_mem():
...
+def broadcast_to_devices():
+    ''' Broadcast cupy ndarray to all the devices, return a list of cupy ndarray
+    '''
+
+
+def reduce_to_device(array_list):
+    ''' Reduce a list of ndarray in different devices to device 0
+    TODO: reduce memory footprint, improve throughput
+    '''
+    num_devices = len(array_list)
```
Comment (sunqm) [inline gpu4pyscf/lib/cupy_helper.py:95]: * early return for single device
* Assert len(array_list) == cudaGetDeviceCount

File: gpu4pyscf/df/grad/jk.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import threading
+import cupy
+from gpu4pyscf.lib.cupy_helper import contract
+from gpu4pyscf.lib import logger
+from gpu4pyscf.__config__ import _streams
+
+def _jk_task(with_df, dm, orbo, device_id, rhoj_total, rhok_total, with_j=True, with_k=True):
```
Comment (sunqm) [inline gpu4pyscf/df/grad/jk.py:22]: default value for `device_id`

File: gpu4pyscf/df/grad/jk.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    return
+
+def get_rhoj_rhok(with_df, dm, orbo, with_j=True, with_k=True):
+    ''' Calculate rhoj and rhok on Multi-GPU system
+    '''
+    num_gpus = cupy.cuda.runtime.getDeviceCount()
+    rhoj_total = [None] * num_gpus
+    rhok_total = [None] * num_gpus
+    threads = []
+    for device_id in range(num_gpus):
```
Comment (sunqm) [inline gpu4pyscf/df/grad/jk.py:66]: `ThreadPoolExecutor`

File: gpu4pyscf/lib/cupy_helper.py [hunk 5]
```diff
@@ -82,6 +83,51 @@ def get_avail_mem():
         mem_avail = cupy.cuda.runtime.memGetInfo()[0]
         return mem_avail + total_mem - used_mem
 
+def broadcast_to_devices():
+    ''' Broadcast cupy ndarray to all the devices, return a list of cupy ndarray
+    '''
```
Comment (sunqm) [inline gpu4pyscf/lib/cupy_helper.py:88]: Raise NotImplementedError

File: gpu4pyscf/mp/dfmp2.py [hunk 6]
```diff
@@ -13,18 +13,49 @@
...
 from gpu4pyscf.lib import logger
-from gpu4pyscf.lib.cupy_helper import contract, tag_array
+from gpu4pyscf.lib.cupy_helper import contract, tag_array, reduce_to_device
+from gpu4pyscf.__config__ import _streams, _num_devices
 from pyscf import __config__
 
 WITH_T2 = getattr(__config__, 'mp_dfmp2_with_t2', True)
 _einsum = cupy.einsum
 
+def _dfmp2_tasks(mp, mo_coeff, mo_energy, device_id, Lov_dist):
```
Comment (sunqm) [inline gpu4pyscf/mp/dfmp2.py:30]: default `device_id`

File: gpu4pyscf/mp/dfmp2.py [hunk 7]
```diff
@@ -37,31 +68,36 @@ def kernel(mp, mo_energy=None, mo_coeff=None, eris=None, with_t2=WITH_T2,
...
     mo_energy = cupy.asarray(mo_energy)
     mo_coeff = cupy.asarray(mo_coeff)
+    
+    if mp.with_df.naux is None:
+        mp.with_df.build()
+
+    Lov_dist = [None] * _num_devices
+    threads = []
+    for device_id in range(_num_devices):
+        thread = threading.Thread(
```
Comment (sunqm) [inline gpu4pyscf/mp/dfmp2.py:78]: `ThreadPoolExecutor`

File: gpu4pyscf/df/df.py [hunk 8]
```diff
@@ -198,103 +199,164 @@ def reset(self, mol=None):
...
     avail_mem = get_avail_mem()
-    use_gpu_memory = True
-    if naux * npair * 8 < 0.4 * avail_mem:
-        try:
-            cderi = cupy.empty([naux, npair], order='C')
-            log.debug(f"Saving CDERI on GPU. CDERI size {cderi.nbytes/GB}")
-        except Exception:
-            use_gpu_memory = False
+    
+    if use_gpu_memory:
```
Comment (sunqm) [inline gpu4pyscf/df/df.py:217]: Should the `num_gpus` be considered when estimating GPU memory capacity?

File: gpu4pyscf/df/df.py [hunk 9]
```diff
@@ -198,103 +199,164 @@ def reset(self, mol=None):
...
+    for gpu_id in range(num_gpus):
+        cupy.cuda.Device(gpu_id).synchronize()
 
     if not use_gpu_memory:
         cupy.cuda.Device().synchronize()
-    return cderi
+    
+    return _cderi
 
+def _cderi_task(intopt, cd_low, task_list, device_id, _cderi, omega=None, sr_only=False):
```
Comment (sunqm) [inline gpu4pyscf/df/df.py:275]: default `device_id`

File: gpu4pyscf/df/df.py (hunk 8 above)
Comment (wxj6000) [reply gpu4pyscf/df/df.py:217]: Yes, there is a todo in the comment. Will do in next PR.

**Comment by wxj6000** [comment]
@sunqm Ready for review.

File: gpu4pyscf/df/df.py [hunk 10]
```diff
@@ -240,47 +238,41 @@ def cholesky_eri_gpu(intopt, mol, auxmol, cd_low,
...
-                                  kwargs={"omega":omega, "sr_only":sr_only})
-        thread.start()
-        threads.append(thread)
+    futures = []
+    with ThreadPoolExecutor(max_workers=_num_devices) as executor:
+        for device_id in range(_num_devices):
+            task_list = task_list_per_device[device_id]
+            future = executor.submit(_cderi_task, intopt, cd_low_f, task_list, _cderi,
+                                     omega=omega, sr_only=sr_only, device_id=device_id)
+            futures.append(future)
```
Comment (sunqm) [inline gpu4pyscf/df/df.py:256]: calls `future.result()` for each Future object. Exceptions in threads, if any,  are raised in this thread synchronization operation. This checks the. The same for the other Thread executors.

**Review by sunqm — CHANGES_REQUESTED** [review]

**Review by sunqm — APPROVED** [review]

**Pull request merged**